#!/bin/bash

AccumulateQualityYieldMetrics () { picard AccumulateQualityYieldMetrics "$@" }
AccumulateVariantCallingMetrics () { picard AccumulateVariantCallingMetrics "$@" }
AddCommentsToBam () { picard AddCommentsToBam "$@" }
AddOATag () { picard AddOATag "$@" }
AddOrReplaceReadGroups () { picard AddOrReplaceReadGroups "$@" }
BaitDesigner () { picard BaitDesigner "$@" }
BamIndexStats () { picard BamIndexStats "$@" }
BamToBfq () { picard BamToBfq "$@" }
BedToIntervalList () { picard BedToIntervalList "$@" }
BpmToNormalizationManifestCsv () { picard BpmToNormalizationManifestCsv "$@" }
BuildBamIndex () { picard BuildBamIndex "$@" }
CalculateFingerprintMetrics () { picard CalculateFingerprintMetrics "$@" }
CalculateReadGroupChecksum () { picard CalculateReadGroupChecksum "$@" }
CheckDuplicateMarking () { picard CheckDuplicateMarking "$@" }
CheckFingerprint () { picard CheckFingerprint "$@" }
CheckIlluminaDirectory () { picard CheckIlluminaDirectory "$@" }
CheckTerminatorBlock () { picard CheckTerminatorBlock "$@" }
CleanSam () { picard CleanSam "$@" }
ClusterCrosscheckMetrics () { picard ClusterCrosscheckMetrics "$@" }
CollectAlignmentSummaryMetrics () { picard CollectAlignmentSummaryMetrics "$@" }
CollectArraysVariantCallingMetrics () { picard CollectArraysVariantCallingMetrics "$@" }
CollectBaseDistributionByCycle () { picard CollectBaseDistributionByCycle "$@" }
CollectDuplicateMetrics () { picard CollectDuplicateMetrics "$@" }
CollectGcBiasMetrics () { picard CollectGcBiasMetrics "$@" }
CollectHiSeqXPfFailMetrics () { picard CollectHiSeqXPfFailMetrics "$@" }
CollectHsMetrics () { picard CollectHsMetrics "$@" }
CollectIlluminaBasecallingMetrics () { picard CollectIlluminaBasecallingMetrics "$@" }
CollectIlluminaLaneMetrics () { picard CollectIlluminaLaneMetrics "$@" }
CollectIndependentReplicateMetrics () { picard CollectIndependentReplicateMetrics "$@" }
CollectInsertSizeMetrics () { picard CollectInsertSizeMetrics "$@" }
CollectJumpingLibraryMetrics () { picard CollectJumpingLibraryMetrics "$@" }
CollectMultipleMetrics () { picard CollectMultipleMetrics "$@" }
CollectOxoGMetrics () { picard CollectOxoGMetrics "$@" }
CollectQualityYieldMetrics () { picard CollectQualityYieldMetrics "$@" }
CollectRawWgsMetrics () { picard CollectRawWgsMetrics "$@" }
CollectRnaSeqMetrics () { picard CollectRnaSeqMetrics "$@" }
CollectRrbsMetrics () { picard CollectRrbsMetrics "$@" }
CollectSamErrorMetrics () { picard CollectSamErrorMetrics "$@" }
CollectSequencingArtifactMetrics () { picard CollectSequencingArtifactMetrics "$@" }
CollectTargetedPcrMetrics () { picard CollectTargetedPcrMetrics "$@" }
CollectVariantCallingMetrics () { picard CollectVariantCallingMetrics "$@" }
CollectWgsMetrics () { picard CollectWgsMetrics "$@" }
CollectWgsMetricsWithNonZeroCoverage () { picard CollectWgsMetricsWithNonZeroCoverage "$@" }
CombineGenotypingArrayVcfs () { picard CombineGenotypingArrayVcfs "$@" }
CompareGtcFiles () { picard CompareGtcFiles "$@" }
CompareMetrics () { picard CompareMetrics "$@" }
CompareSAMs () { picard CompareSAMs "$@" }
ConvertHaplotypeDatabaseToVcf () { picard ConvertHaplotypeDatabaseToVcf "$@" }
ConvertSequencingArtifactToOxoG () { picard ConvertSequencingArtifactToOxoG "$@" }
CreateBafRegressMetricsFile () { picard CreateBafRegressMetricsFile "$@" }
CreateExtendedIlluminaManifest () { picard CreateExtendedIlluminaManifest "$@" }
CreateSequenceDictionary () { picard CreateSequenceDictionary "$@" }
CreateVerifyIDIntensityContaminationMetricsFile () { picard CreateVerifyIDIntensityContaminationMetricsFile "$@" }
CrosscheckFingerprints () { picard CrosscheckFingerprints "$@" }
CrosscheckReadGroupFingerprints () { picard CrosscheckReadGroupFingerprints "$@" }
DownsampleSam () { picard DownsampleSam "$@" }
EstimateLibraryComplexity () { picard EstimateLibraryComplexity "$@" }
ExtractFingerprint () { picard ExtractFingerprint "$@" }
ExtractIlluminaBarcodes () { picard ExtractIlluminaBarcodes "$@" }
ExtractSequences () { picard ExtractSequences "$@" }
FastqToSam () { picard FastqToSam "$@" }
FifoBuffer () { picard FifoBuffer "$@" }
FilterSamReads () { picard FilterSamReads "$@" }
FilterVcf () { picard FilterVcf "$@" }
FindMendelianViolations () { picard FindMendelianViolations "$@" }
FixMateInformation () { picard FixMateInformation "$@" }
FixVcfHeader () { picard FixVcfHeader "$@" }
GatherBamFiles () { picard GatherBamFiles "$@" }
GatherVcfs () { picard GatherVcfs "$@" }
GenotypeConcordance () { picard GenotypeConcordance "$@" }
GtcToVcf () { picard GtcToVcf "$@" }
IdentifyContaminant () { picard IdentifyContaminant "$@" }
IlluminaBasecallsToFastq () { picard IlluminaBasecallsToFastq "$@" }
IlluminaBasecallsToSam () { picard IlluminaBasecallsToSam "$@" }
IntervalListToBed () { picard IntervalListToBed "$@" }
IntervalListTools () { picard IntervalListTools "$@" }
LiftOverHaplotypeMap () { picard LiftOverHaplotypeMap "$@" }
LiftOverIntervalList () { picard LiftOverIntervalList "$@" }
LiftoverVcf () { picard LiftoverVcf "$@" }
MakeSitesOnlyVcf () { picard MakeSitesOnlyVcf "$@" }
MakeVcfSampleNameMap () { picard MakeVcfSampleNameMap "$@" }
MarkDuplicates () { picard MarkDuplicates "$@" }
MarkDuplicatesWithMateCigar () { picard MarkDuplicatesWithMateCigar "$@" }
MarkIlluminaAdapters () { picard MarkIlluminaAdapters "$@" }
MeanQualityByCycle () { picard MeanQualityByCycle "$@" }
MergeBamAlignment () { picard MergeBamAlignment "$@" }
MergePedIntoVcf () { picard MergePedIntoVcf "$@" }
MergeSamFiles () { picard MergeSamFiles "$@" }
MergeVcfs () { picard MergeVcfs "$@" }
NonNFastaSize () { picard NonNFastaSize "$@" }
NormalizeFasta () { picard NormalizeFasta "$@" }
PositionBasedDownsampleSam () { picard PositionBasedDownsampleSam "$@" }
QualityScoreDistribution () { picard QualityScoreDistribution "$@" }
RenameSampleInVcf () { picard RenameSampleInVcf "$@" }
ReorderSam () { picard ReorderSam "$@" }
ReplaceSamHeader () { picard ReplaceSamHeader "$@" }
RevertOriginalBaseQualitiesAndAddMateCigar () { picard RevertOriginalBaseQualitiesAndAddMateCigar "$@" }
RevertSam () { picard RevertSam "$@" }
SamFormatConverter () { picard SamFormatConverter "$@" }
SamToFastq () { picard SamToFastq "$@" }
SamToFastqWithTags () { picard SamToFastqWithTags "$@" }
ScatterIntervalsByNs () { picard ScatterIntervalsByNs "$@" }
SetNmAndUqTags () { picard SetNmAndUqTags "$@" }
SetNmMdAndUqTags () { picard SetNmMdAndUqTags "$@" }
SimpleMarkDuplicatesWithMateCigar () { picard SimpleMarkDuplicatesWithMateCigar "$@" }
SortGff () { picard SortGff "$@" }
SortSam () { picard SortSam "$@" }
SortVcf () { picard SortVcf "$@" }
SplitSamByLibrary () { picard SplitSamByLibrary "$@" }
SplitSamByNumberOfReads () { picard SplitSamByNumberOfReads "$@" }
SplitVcfs () { picard SplitVcfs "$@" }
UmiAwareMarkDuplicatesWithMateCigar () { picard UmiAwareMarkDuplicatesWithMateCigar "$@" }
UpdateVcfSequenceDictionary () { picard UpdateVcfSequenceDictionary "$@" }
ValidateSamFile () { picard ValidateSamFile "$@" }
VcfFormatConverter () { picard VcfFormatConverter "$@" }
VcfToAdpc () { picard VcfToAdpc "$@" }
VcfToIntervalList () { picard VcfToIntervalList "$@" }
ViewSam () { picard ViewSam "$@" }
export -f AccumulateQualityYieldMetrics AccumulateVariantCallingMetrics AddCommentsToBam AddOATag AddOrReplaceReadGroups BaitDesigner BamIndexStats BamToBfq BedToIntervalList BpmToNormalizationManifestCsv BuildBamIndex CalculateFingerprintMetrics CalculateReadGroupChecksum CheckDuplicateMarking CheckFingerprint CheckIlluminaDirectory CheckTerminatorBlock CleanSam ClusterCrosscheckMetrics CollectAlignmentSummaryMetrics CollectArraysVariantCallingMetrics CollectBaseDistributionByCycle CollectDuplicateMetrics CollectGcBiasMetrics CollectHiSeqXPfFailMetrics CollectHsMetrics CollectIlluminaBasecallingMetrics CollectIlluminaLaneMetrics CollectIndependentReplicateMetrics CollectInsertSizeMetrics CollectJumpingLibraryMetrics CollectMultipleMetrics CollectOxoGMetrics CollectQualityYieldMetrics CollectRawWgsMetrics CollectRnaSeqMetrics CollectRrbsMetrics CollectSamErrorMetrics CollectSequencingArtifactMetrics CollectTargetedPcrMetrics CollectVariantCallingMetrics CollectWgsMetrics CollectWgsMetricsWithNonZeroCoverage CombineGenotypingArrayVcfs CompareGtcFiles CompareMetrics CompareSAMs ConvertHaplotypeDatabaseToVcf ConvertSequencingArtifactToOxoG CreateBafRegressMetricsFile CreateExtendedIlluminaManifest CreateSequenceDictionary CreateVerifyIDIntensityContaminationMetricsFile CrosscheckFingerprints CrosscheckReadGroupFingerprints DownsampleSam EstimateLibraryComplexity ExtractFingerprint ExtractIlluminaBarcodes ExtractSequences FastqToSam FifoBuffer FilterSamReads FilterVcf FindMendelianViolations FixMateInformation FixVcfHeader GatherBamFiles GatherVcfs GenotypeConcordance GtcToVcf IdentifyContaminant IlluminaBasecallsToFastq IlluminaBasecallsToSam IntervalListToBed IntervalListTools LiftOverHaplotypeMap LiftOverIntervalList LiftoverVcf MakeSitesOnlyVcf MakeVcfSampleNameMap MarkDuplicates MarkDuplicatesWithMateCigar MarkIlluminaAdapters MeanQualityByCycle MergeBamAlignment MergePedIntoVcf MergeSamFiles MergeVcfs NonNFastaSize NormalizeFasta PositionBasedDownsampleSam QualityScoreDistribution RenameSampleInVcf ReorderSam ReplaceSamHeader RevertOriginalBaseQualitiesAndAddMateCigar RevertSam SamFormatConverter SamToFastq SamToFastqWithTags ScatterIntervalsByNs SetNmAndUqTags SetNmMdAndUqTags SimpleMarkDuplicatesWithMateCigar SortGff SortSam SortVcf SplitSamByLibrary SplitSamByNumberOfReads SplitVcfs UmiAwareMarkDuplicatesWithMateCigar UpdateVcfSequenceDictionary ValidateSamFile VcfFormatConverter VcfToAdpc VcfToIntervalList ViewSam
bash -c "$@"